File: backend/api/database/book.py
from api.database.db import db
from sqlalchemy.orm import Mapped, mapped_column
from datetime import datetime
from decimal import Decimal
from uuid import UUID


class Book(db.Model):
    __tablename__ = "book"

    id: Mapped[UUID] = mapped_column(primary_key=True, nullable=False)
    title: Mapped[str] = mapped_column(nullable=False)
    author: Mapped[str] = mapped_column(nullable=False)
    genre: Mapped[str] = mapped_column(nullable=False)
    description: Mapped[str] = mapped_column(nullable=True)
    rating: Mapped[Decimal] = mapped_column(nullable=True)
    date_published: Mapped[datetime] = mapped_column(nullable=False)

    def to_dict(self):
        return {
            "id": str(self.id),
            "title": self.title,
            "author": self.author,
            "genre": self.genre,
            "description": self.description,
            "rating": float(self.rating),
            "date_published": str(self.date_published)
        }

File: backend/src/model/book.rs
use core::fmt;

use sqlx::FromRow;
use serde::{Deserialize, Serialize};
use chrono::{DateTime, Utc};  // Utc for UTC times
use uuid::Uuid;  // Import the Uuid type
use rust_decimal::Decimal;

#[derive(Debug, Serialize, Deserialize, FromRow)]
pub struct Book{
    pub id: Uuid,
    pub title: String,
    pub author: String,
    pub genre: String,
    pub description: Option<String>,
    pub rating: Option<Decimal>,
    pub date_published: Option<DateTime<Utc>>,
}

#[derive(Debug, Serialize, Deserialize)]

pub struct CreateBookRequest{
    pub title: String,
    pub author: String,
    pub genre: String,
    pub description: Option<String>,
    pub rating: Option<Decimal>,
    pub date_published: Option<DateTime<Utc>>,
}

#[derive(Debug, Serialize, Deserialize)]

pub struct UpdateBookRequest{
    pub title: Option<String>,
    pub author: Option<String>,
    pub genre: Option<String>,
    pub description: Option<String>,
    pub rating: Option<Decimal>,
    pub date_published: Option<DateTime<Utc>>,
}


impl fmt::Display for Book{
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "({}, {}, {}, {}, {:?}, {:?}, {:?})", self.id, self.title, self.author, self.genre, self.description, self.rating, self.date_published)
    }
}


impl fmt::Display for CreateBookRequest{
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "({}, {}, {}, {:?}, {:?}, {:?})", self.title, self.author, self.genre, self.description, self.rating, self.date_published)
    }
}

impl fmt::Display for UpdateBookRequest{
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "({:?}, {:?}, {:?}, {:?}, {:?}, {:?})", self.title, self.author, self.genre, self.description, self.rating, self.date_published)
    }
}